### ISSUE TITLE
tf.tf.Variable() seems cannot be replaced with tf.get_variable().

### ISSUE THREAD

**Issue by nooooonee** [MadryLab/mnist_challenge#9]
I just replace the  `tf.Variable()` in function `_weight_variable`  and `_bias_variable` with `tf.get_variable()`. And it cannot train a robust network to resist CW attack.  In contrast, I run the unchanged source code, it can train a robust network, I am really confused why it is? The following are the code I only changed. Please help. 

  

  ```
@staticmethod
  def _weight_variable(shape, name):
      initial = tf.initializers.truncated_normal(stddev=0.1)
      return tf.get_variable(shape=shape, name=name, initializer=initial)
@staticmethod
def _bias_variable(shape, name):
      initial = tf.constant(0.1, shape=shape)
      return tf.get_variable(name=name, initializer=initial)
```

**Comment by dtsip** [comment]
This is weird. Looks like an issue with tensorflow though, so I don't see how we can help here.

**Comment by nooooonee** [comment]
Later, I am thinking is adversarial training with PGD not necessarily robust to CW attack (I mean original CW attack, not PGD attack with CW loss function in your paper)? Is PGD-based adversarial trained network robust to CW  attack just an accidental phenomenon under different initialization settings?

**Comment by dtsip** [comment]
The goal of PGD training is to solve the min-max problem stated in our paper. If it is successful, there is _no_ attack that will degrade the accuracy of our model (be it standard CW or whatever variant). In fact, we have found that PGD training leads to models that can be _provably_ robust (https://arxiv.org/abs/1809.03008).

**Comment by nooooonee** [comment]
So because pgd training doesn't completely solve the  min-max problem，it can not ensure target network CW to be robust to CW or other perturbation-based adversarial examples（of course，it is robust to pgd attack ）. And if we can globally resolve the min-max problem，we will be able to resist all perturbation based adversarial example. Just to make it clear，do you mean these？

获取 Outlook for iOS<https://aka.ms/o0ukef>

________________________________
发件人: Dimitris Tsipras <notifications@github.com>
发送时间: 星期一, 二月 25, 2019 03:31
收件人: MadryLab/mnist_challenge
抄送: lepangdan; Author
主题: Re: [MadryLab/mnist_challenge] tf.tf.Variable() seems cannot be replaced with tf.get_variable(). (#9)


The goal of PGD training is to solve the min-max problem stated in our paper. If it is successful, there is no attack that will degrade the accuracy of our model (be it standard CW or whatever variant). In fact, we have found that PGD training leads to models that can be provably robust (https://arxiv.org/abs/1809.03008).

―

**Comment by dtsip** [comment]
PGD _does_ solve the min-max problem well enough. As I mentioned before, PGD trained networks have been shown to be _provably_ robust to _every_ adversarial attack within the threat model (https://arxiv.org/abs/1809.03008).

**Comment by dtsip** [comment]
Moreover, the `secret` network in our challenge _is_ robust to the standard CW attack. And I wouldn't call this an accident.

**Comment by nooooonee** [comment]
okay，thank you so much，i will follow the paper you recommend. Yes，i found both the released adv_train and secret models are robust to cw，and also because this，i am confused for long time why i can not train robust network to cw，and i even think is it a bug of tf.get_variable()  function，this why i open the issue. Thanks a lot for let me know PGD training in your code can not always defend cw, which is important for me.

获取 Outlook for iOS<https://aka.ms/o0ukef>

________________________________
发件人: Dimitris Tsipras <notifications@github.com>
发送时间: 星期一, 二月 25, 2019 03:50
收件人: MadryLab/mnist_challenge
抄送: lepangdan; Author
主题: Re: [MadryLab/mnist_challenge] tf.tf.Variable() seems cannot be replaced with tf.get_variable(). (#9)


Moreover, the secret network in our challenge is robust to the standard CW attack. And I wouldn't call this an accident.

―

**Comment by lepangdan** [comment]
Sorry，please ignore the last mail，i seem to misunderstand what you mean.  Then， we key question i want to know is if we can steadily train a network robust to cw pgd training in your code.

获取 Outlook for iOS<https://aka.ms/o0ukef>

________________________________
发件人: pang dan <lepangdan@outlook.com>
发送时间: 星期一, 二月 25, 2019 04:02
收件人: MadryLab/mnist_challenge; MadryLab/mnist_challenge
抄送: Author
主题: Re: [MadryLab/mnist_challenge] tf.tf.Variable() seems cannot be replaced with tf.get_variable(). (#9)

okay，thank you so much，i will follow the paper you recommend. Yes，i found both the released adv_train and secret models are robust to cw，and also because this，i am confused for long time why i can not train robust network to cw，and i even think is it a bug of tf.get_variable()  function，this why i open the issue. Thanks a lot for let me know PGD training in your code can not always defend cw, which is important for me.

获取 Outlook for iOS<https://aka.ms/o0ukef>

________________________________
发件人: Dimitris Tsipras <notifications@github.com>
发送时间: 星期一, 二月 25, 2019 03:50
收件人: MadryLab/mnist_challenge
抄送: lepangdan; Author
主题: Re: [MadryLab/mnist_challenge] tf.tf.Variable() seems cannot be replaced with tf.get_variable(). (#9)


Moreover, the secret network in our challenge is robust to the standard CW attack. And I wouldn't call this an accident.

―

**Comment by nooooonee** [comment]
Sorry，please ignore the last mail，i seem to misunderstand what you mean. Then， the key question i want to know is if i can steadily train a network robust to cw in your code.

获取 Outlook for iOS<https://aka.ms/o0ukef>

________________________________
发件人: pang dan <lepangdan@outlook.com>
发送时间: 星期一, 二月 25, 2019 04:09
收件人: MadryLab/mnist_challenge; MadryLab/mnist_challenge
抄送: Author
主题: Re: [MadryLab/mnist_challenge] tf.tf.Variable() seems cannot be replaced with tf.get_variable(). (#9)

Sorry，please ignore the last mail，i seem to misunderstand what you mean.  Then， we key question i want to know is if we can steadily train a network robust to cw pgd training in your code.

获取 Outlook for iOS<https://aka.ms/o0ukef>

________________________________
发件人: pang dan <lepangdan@outlook.com>
发送时间: 星期一, 二月 25, 2019 04:02
收件人: MadryLab/mnist_challenge; MadryLab/mnist_challenge
抄送: Author
主题: Re: [MadryLab/mnist_challenge] tf.tf.Variable() seems cannot be replaced with tf.get_variable(). (#9)

okay，thank you so much，i will follow the paper you recommend. Yes，i found both the released adv_train and secret models are robust to cw，and also because this，i am confused for long time why i can not train robust network to cw，and i even think is it a bug of tf.get_variable()  function，this why i open the issue. Thanks a lot for let me know PGD training in your code can not always defend cw, which is important for me.

获取 Outlook for iOS<https://aka.ms/o0ukef>

________________________________
发件人: Dimitris Tsipras <notifications@github.com>
发送时间: 星期一, 二月 25, 2019 03:50
收件人: MadryLab/mnist_challenge
抄送: lepangdan; Author
主题: Re: [MadryLab/mnist_challenge] tf.tf.Variable() seems cannot be replaced with tf.get_variable(). (#9)


Moreover, the secret network in our challenge is robust to the standard CW attack. And I wouldn't call this an accident.

―
